packages/snaps-sdk/src/internals/jsx.ts

```diff
@@ -0,0 +1,98 @@
+import type { Infer, Struct } from 'superstruct';
+import type {
+  AnyStruct,
+  EnumSchema,
+  InferStructTuple,
+  IsExactMatch,
+  IsMatch,
+  IsRecord,
+  IsTuple,
+  UnionToIntersection,
+} from 'superstruct/dist/utils';
+
+import type { EmptyObject } from '../types';
+import { union } from './structs';
+
+/**
+ * Check if a type is a union. Infers `true` if it is a union, otherwise
+ * `false`.
+ */
+type IsUnion<Type> = [Type] extends [UnionToIntersection<Type>] ? false : true;
+
+/**
+ * Get a struct schema for a type.
+ *
+ * This is copied from `superstruct` but fixes some issues with the original
+ * implementation.
+ */
+type StructSchema<Type> = IsUnion<Type> extends true
```

Comment:
ouch

Reply:
It's not pretty, but it ensures our validation matches our types, so it's a worthy tradeoff in my opinion.